src/ws/cert-session.c

```diff
@@ -0,0 +1,307 @@
+  assert (certificate);
+
+  user = sssd_map_certificate (certificate);
+  /* should already exit_init_problem() with meaningful error code */
+  assert (user);
+  res = pam_start ("cockpit", user, &conv, &pamh);
+  if (res != PAM_SUCCESS)
+    errx (EX, "couldn't start pam: %s", pam_strerror (NULL, res));
+
+  if (pam_set_item (pamh, PAM_RHOST, rhost) != PAM_SUCCESS)
+    errx (EX, "couldn't setup pam");
+
+  /* no pam_authenticate here, the sssd query replaces the "auth" stage. The
+   * "session" stage still runs via open_session() */
+
+  res = open_session (pamh);
+  /* Our exit code is a PAM code */
+  if (res != PAM_SUCCESS)
+    exit_init_problem (res);
+
+  return pamh;
+}
+
+static int
+session (char **env)
+{
+  char *argv[] = { "cockpit-bridge", NULL };
+  debug ("executing bridge: %s", argv[0]);
+
+  if (env)
+    execvpe (argv[0], argv, env);
+  else
+    execvp (argv[0], argv);
+
+  warn ("can't exec %s", argv[0]);
+  return EX;
+}
+
+int
+main (int argc,
+      char **argv)
+{
+  pam_handle_t *pamh = NULL;
+  const char *rhost;
+  char *authorization;
+  char *type = NULL;
+  const char *auth_value;
+  char **env;
+  int status;
+  int res;
+  int i;
+
+  program_name = basename (argv[0]);
+
+  if (isatty (0))
+    errx (2, "this command is not meant to be run from the console");
+
+  if (argc != 2)
+    errx (2, "invalid arguments to cockpit-cert-session");
+
+  /* Cleanup the umask */
+  umask (077);
+
+  rhost = getenv ("COCKPIT_REMOTE_PEER") ?: "";
+
+  save_environment ();
+
+  /* When setuid root, make sure our group is also root */
+  if (geteuid () == 0)
+    {
+      /* Always clear the environment */
+      if (clearenv () != 0)
+        err (1, "couldn't clear environment");
+
+      /* set a minimal environment */
+      setenv ("PATH", DEFAULT_PATH, 1);
+
+      if (setgid (0) != 0 || setuid (0) != 0)
+        err (1, "couldn't switch permissions correctly");
+    }
+
+  signal (SIGALRM, SIG_DFL);
+  signal (SIGQUIT, SIG_DFL);
+  signal (SIGTSTP, SIG_IGN);
+  signal (SIGHUP, SIG_IGN);
+  signal (SIGPIPE, SIG_IGN);
+
+  cockpit_authorize_logger (authorize_logger, DEBUG_SESSION);
+
+  /* Request authorization header */
+  write_authorize_begin ();
+  write_control_string ("challenge", "*");
+  write_control_end ();
+
+  /* And get back the authorization header */
+  authorization = read_authorize_response ("authorization");
+  auth_value = cockpit_authorize_type (authorization, &type);
+  if (!auth_value)
+    errx (EX, "invalid authorization header received");
+  if (strcmp (type, "tls-cert") != 0)
+    errx (2, "This authentication command only supports tls-cert");
+
+  pamh = perform_tlscert (rhost, auth_value);
+
+  cockpit_memory_clear (authorization, -1);
+  free (authorization);
+  free (type);
+
+  for (i = 0; env_saved[i] != NULL; i++)
+    pam_putenv (pamh, env_saved[i]);
+
+  env = pam_getenvlist (pamh);
+  if (env == NULL)
+    errx (EX, "get pam environment failed");
+
+  if (want_session)
```

Comment:
Where is this defined? I think think cockpit-cert-session doesn't even need want_session == FALSE.

Reply:
It's common code between cockpit-session and cockpit-cert-session, defined in src/ws/session-utils.c. In short, want_session is FALSE if you start cockpit-cert-session as the target user. This is really mostly useful for testing (session.c does exactly the same), but if you find that dubious I'm happy to kill it.